Sprint 6

v-alekseev opened the pull request.

v-alekseev committed:
Текст в StatusBar сделал белым

v-alekseev committed:
alertViewController теперь создается во   viewDidLoad

v-alekseev committed:
добавлена  загрузка из сети в рамках требований  спринта 6

v-alekseev committed:
исправлена работа с индикатором загрузки

v-alekseev committed:
и еще небольшой фикс показа лоадера

Helvetii951 commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -167,29 +166,85 @@ final class MovieQuizViewController: UIViewController, QuestionFactoryDelegate {
         print(log + " File: \(#file) function: \(#function), line: \(#line)")
     }
     
+    private func showLoadingIndicator() {
+        activityIndicator.isHidden = false // говорим, что индикатор загрузки не скрыт
+        activityIndicator.startAnimating() // включаем анимацию
+    }
+    
+    private func hideLoadingIndicator() {
+        activityIndicator.stopAnimating() // выключаем анимацию
+        activityIndicator.isHidden = true // говорим, что индикатор загрузки скрыт
+    }
```
Рекомендую посмотреть на свойство [hidesWhenStopped](https://developer.apple.com/documentation/uikit/uiactivityindicatorview/1622837-hideswhenstopped) 
у `activityIndicator`, и использовать его вместо ручного управления свойством `isHidden`. 

Когда ты будешь вызывать `startAnimating` – индикатор будет появляться автоматически, соответственно на `stopAnimating` – скрываться, а свойство `isHidden` больше не понадобится, индикатор все сделает за тебя. 

**Важно:** это свойство настраивают всего раз раз: кодом во `viewDidLoad()` или аналогичной галочкой у элемента на сториборде.
А строки `activityIndicator.isHidden = false(true)` нужно будет убрать

v-alekseev replied:
готово

Helvetii951 commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -167,29 +166,85 @@ final class MovieQuizViewController: UIViewController, QuestionFactoryDelegate {
         }
     }
     
+    func didLoadDataFromServer() {
+        //скрываем индикатор загрузки в didReceiveNextQuestion т/к/ последует долгая загрузка картинок.
+        questionFactory?.requestNextQuestion()
+    }
```
Ты можешь дополнительно показывать `activityIndicator` между загрузкой вопросов, потому что загрузка вопроса тоже занимает некоторое время из-за получения изображения.
Перед запросом следующего вопроса – показывать индикатор, а в момент получения – выключать анимацию.

Helvetii951 commented on MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -10,39 +10,84 @@ import Foundation
+            //self.delegate?.showLoadingIndicator()
+            do {
+                imageData = try Data(contentsOf: movie.resizedImageURL)
+            } catch {
+                print("Failed to load image")
             }
```
Обращу внимание, что игра состоит из вопроса о рейтинге, на который отвечает пользователь, и контекста о фильме в виде постера(картинка). 
Без постера, пользователю будет затруднительно ответить на вопрос о рейтинге фильма верно.

В этом месте, функция не завершает свою работу и идет формировать вопрос без постера. 
В таком случае, бизнес-логика приложения будет нарушена, потому что пользователь не сможет ответить корректно, ведь изображения нет.
Это состояние стоит обработать, например, с помощью алерта. 
Не забудь, что алерт будет обновлять UI и это тоже нужно учесть (вернуть на main), ровно также, как и прерывание функции (сделать return)

Helvetii951 commented on MovieQuiz/Presentation/AlertPresenter.swift:
```diff
@@ -9,11 +9,11 @@ import UIKit
-    init(parentViewController: UIViewController){
-        self.parentViewController = parentViewController
-    }
+//    init(parentViewController: UIViewController){
+//        self.parentViewController = parentViewController
+//    }
```
Рекомендую убирать комментарии устаревшего кода. Будет выглядеть чище, легче будет читать.

Helvetii951 commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -167,29 +166,85 @@ final class MovieQuizViewController: UIViewController, QuestionFactoryDelegate {
     // MARK: - Lifecycle
     override func viewDidLoad() {
         super.viewDidLoad()
+        // фактически инициализируем класс показа алерта
```
И такие технические комментарии тоже лучше убирать.

v-alekseev replied:
убран

Helvetii951 commented on MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -10,39 +10,84 @@ import Foundation
 final class QuestionFactory: QuestionFactoryProtocol {
     private var currentIndex = 0
     private var previosIndex = 0
+    private let ratingLevel: Double = 8.3
```
Думаю, предполагаемый рейтинг формируется слишком просто :) Попробуй сделать его рандомным. Например в интервале 6...9.

v-alekseev committed:
- рамка создается только в viewDidLoad()
- если не прошла начальная загрузка или не загрузилась картинка, нужно показывать белый фрейм постера. Вроде этого - скрин примера.

v-alekseev committed:
- hidesWhenStopped поставлено в true
- Рейтиг в вопрос берется случайный (75...91).randomElement()
- обрабатывается ошибка загрузки картинки при генерации вопроса. Покаываем Алерт и предлагаем звгрузить следующий вопрос

v-alekseev committed:
убран старый комментарий
